### PR TITLE
Version 3: Implement G element (maximum)

### PR DESCRIPTION
**Pull request by Lyxal**

### CODE REVIEW

**Comment by GingerIndustries** [comment]
nice number

**Review by nayakrujul — APPROVED** [review]

**Review by Steffan153 — APPROVED** [review]

File: shared/src/main/scala/Elements.scala [hunk 1]
```diff
@@ -524,6 +524,34 @@ object Elements:
...
+    val maximum = addDirect(
+      "G",
+      "Monadic Maximum | Dyadic Maximum | Generate From Function | Vectorised Maximum",
+      List("max", "maximum", "generator"),
+      Some(2),
+      "a: lst -> Maximum of a",
+      "a: non-lst, b: non-lst -> Maximum of a and b",
+      "a: lst, b: fun -> Call b infinitely with items of a as starting values"
+    ) { ctx ?=>
+      val top = ctx.pop();
```
Comment (Steffan153) [inline shared/src/main/scala/Elements.scala:536]:
```suggestion
      val top = ctx.pop()
```

File: shared/src/main/scala/ListHelpers.scala [hunk 2]
```diff
@@ -107,6 +130,14 @@ object ListHelpers:
 
   end map
 
+  def maximum(iterable: VList)(using ctx: Context): VAny =
+    iterable match
+      case VList() => VList()
+      case VList(list*) =>
+        list.reduce { (a, b) =>
+          if MiscHelpers.compareExact(a, b) > 0 then a else b
+        }
```
Comment (Steffan153) [inline shared/src/main/scala/ListHelpers.scala:139]: Can you do something like
```
list.min(Ordering.fromLessThan[(VAny, VAny)](MiscHelpers.compareExact(_, _)))
```

File: shared/src/main/scala/ListHelpers.scala [hunk 3]
```diff
@@ -211,4 +242,11 @@ object ListHelpers:
 
     parts.toSeq
   end split
+
+  def vectorisedMaximum(iterable: VList, b: VVal): VList =
+    VList.from(iterable.map { a =>
+      (a: @unchecked) match
+        case a: VList => vectorisedMaximum(a, b)
+        case a: VVal  => MiscHelpers.dyadicMaximum(a, b)
+    })
```
[the remark this answers is not in the document]
Comment (Lyxal) [reply shared/src/main/scala/ListHelpers.scala:250]: `iterable` might contain further lists, which aren't instances of `VVal`

File: shared/src/main/scala/Elements.scala [hunk 4]
```diff
@@ -524,6 +524,34 @@ object Elements:
...
+      top match
+        case a: VList =>
+          ctx.push(ListHelpers.maximum(a))
+        case _ =>
+          val next = ctx.pop()
+          (top, next) match
+            case (a: VVal, b: VVal) => ctx.push(MiscHelpers.dyadicMaximum(a, b))
+            case (a: VList, b: VFun) =>
```
[the remark this answers is not in the document]
Comment (Lyxal) [reply shared/src/main/scala/Elements.scala:544]: You're right. And in testing I had to put the function as top anyway. I blame copilot for making me think reversibility was needed :p

File: shared/src/main/scala/ListHelpers.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (Lyxal) [reply shared/src/main/scala/ListHelpers.scala:250]: I don't know if it would work type-wise

File: shared/src/main/scala/ListHelpers.scala (hunk 2 above)
Comment (Lyxal) [reply shared/src/main/scala/ListHelpers.scala]: Maybe, but I prefer the style of the reduce-by better.

Comment (Steffan153) [reply shared/src/main/scala/ListHelpers.scala]: Yes, but are you sure this isn't faster?

**Pull request merged**